src/combinatorics/young_diagram.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2022 Jake Levinson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jake Levinson
+-/
+import order.upper_lower
+
+/-!
+# Young diagrams
+
+A Young diagram is a finite set of up-left justified boxes:
+
+```text
+□□□□□
+□□□
+□□□
+□
+```
+This Young diagram corresponds to the [5, 3, 3, 1] partition of 12.
+
+We represent it as a lower set in `ℕ × ℕ` in the product partial order. We write `(i, j) ∈ μ`
+to say that `(i, j)` (in matrix coordinates) is in the Young diagram `μ`.
+
+## Main functions:
+
+  - `young_diagram` : basic definition
+  - `young_diagram.distrib_lattice`
+  - `young_diagram.card` : the cardinality (number of boxes)
+
+  - various constructors:
+      `young_diagram.has_bot` : the empty Young diagram is (⊥ : young_diagram)
+
+## Notation:
+
+In "English notation", a Young diagram is drawn so that (i1, j1) ≤ (i2, j2)
+means (i1, j1) is weakly up-and-left (northwest) of (i2, j2). This terminology is used
+below, e.g. in `young_diagram.nw_of`.
+
+## Tags
+
+Young diagram
+
+## References
+
+<https://en.wikipedia.org/wiki/Young_tableau>
+
+-/
+
+section young_diagram
+
+/-- A Young diagram is a lower set in ℕ × ℕ -/
```

Comment:
I haven't checked line lengths here:
```suggestion
/-- A Young diagram is a finite collection of cells on the `ℕ × ℕ` grid such that whenever
a cell is present, so are all the ones above and to the left of it. Like matrices, a `(i, j)` cell
is a cell in row `i` and column `j`, where rows are enumerated downward and
columns rightward.

Young diagrams are modeled as finite sets in `ℕ × ℕ` that are lower sets with respect to the standard order on products. -/
```

Reply:
Will do this one manually. Also, I'd rather put the lengthy description in the module docstring where it can substitute for the text on lines 11-22.

Reply:
There's a point to this documentation duplication -- it makes it easier for someone to come across what the conventions are. It's fine (and good I think) to have extra documentation on definitions, and a handful of lines isn't really lengthy.

Reply:
Another thing is that when you hover over a definition or lemma you get to see the docstring immediately, which can be helpful for understanding some unfamiliar code. But you don't get to see the module docstring without actually navigating to the module.